text/0000-placement-traits.md

```diff
@@ -0,0 +1,312 @@
+    fn make_place(self, layout: Layout) -> Self::Place
+        where Self: Sized,
+    {
+        self.make_place_dynamic(Layout::new::<D>())
+    }
+}
+
+default impl<D, A> Boxer<D, A> for T
+    where T: DstBoxer<D, A>
+          A: Allocator
+{
+    fn make_place(allocator: A) -> Self::Place
+        where Self: Sized,
+    {
+        self.make_place_dynamic(Layout::new::<D>())
+    }
+}
+```
+
+## Choices
+
+### Placer::make_place
+
+`Placer::make_place` takes self by value. Taking self by reference as discussed
+in #1286 would either require an option dance, HKT to properly handle lifetimes
+(`type Place<'a> = ...`), or shenanigans. Furthermore, taking self by value
+ensures that non-copy placers are used only once. This allows panic-free
+allocation using `try!(place) <- thing`.
+
+### Parameterize Placer with Owner
+
+`Owner` is now a type parameter in `Placer`. This allows placement in syntax to
+be used with allocators. That is:
+
+```rust
+let boxed_thing: Box<_> = HEAP <- thing1;
+let rced_thing: Rc<_> = HEAP <- thing2;
+```
+
+If `PLACER <- thing` can have precisely one type as in the original RFC, it
+wouldn't be possible to produce both `Rc`s and `Box`s (see alternatives for an
+alternative solution).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Placer::make_place self by value
+
+Taking self by-value means that `vec <- value` won't work. However, `vec.back()
+<- value` still works just fine so I'm not convinced this is a problem.
+
+The other drawback is that `arena <- value` won't work either. One solution is
+to autoref the placer if necessary but,
+
+1. I'm not entirely convinced this use case is common enough to be worth it.
+2. This is a backwards compatible change that can be made at any time.
+
+## Place taking Owner as a type parameter
+
+This could interfere with type inference but shouldn't be an issue. In most
+cases, `Placer` will only be defined for one `Owner` per type.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Instead of making the `Placer` trait take `Owner` as a type argument, we could
+add the following default method to the `Allocator` trait:
+
+```rust
+trait Allocator {
+    /* ... */
+    fn emplace<B: Boxer>(self) -> BoxPlacer<Self, B> {
+        BoxPlacer {
+            allocator: self,
+            _marker: PhantomData,
+        }
+    }
+}
+
+pub struct BoxPlacer<A: Allocator, B> {
+    allocator: A,
+    _marker: PhantomData(fn() -> B)
+}
+
+impl<A, B, D> Placer<D> for BoxPlacer<A, B>
+    where B: Boxer<D>,
+          A: Allocator,
+{
+    fn make_place(self) -> B::Place
+        where Self: Sized,
+    {
+        <B as Boxer<D>>::make_place(self)
+    }
+}
+```
+
+We could then use `let boxed = HEAP.emplace::<Type<_>>() <- value;` to select
+the output type. However, IMO, this is much less ergonomic.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Does `Place::pointer` need to return an `&mut T` (or some other special pointer)
```

Comment:
I'm pretty sure it would be a mistake to have it return `&mut T`, since you'd have to add all the caveats like "there isn't *actually* a `T` behind the reference, so don't call methods on it ... or swap it out with another `T` ..."